mdtk/degradations.py

```diff
@@ -1 +1,565 @@
 """Code to perform the degradations i.e. edits to the midi data"""
+import numpy as np
+import warnings
+from numpy.random import randint, uniform, choice
+
+from mdtk.data_structures import Composition
+
+
+
+MIN_PITCH = 0
+MAX_PITCH = 127
+
+
+
+def set_random_seed(func, seed=None):
+    """This is a function decorator which just adds the keyword argument `seed`
+    to the end of the supplied function that it decorates. It seeds numpy's
+    random state with the provided value before the call of the function.
+
+    Parameters
+    ----------
+    func : function
+        function to be decorated
+    seed : int or None
+        A seed to be supplied to np.random.seed(). Defaults to None, which
+        leaves numpy's random state unchanged.
+
+    Returns
+    -------
+    seeded_func : function
+        The originally supplied function, but now with an aditional optional
+        seed keyword argument.
+    """
+    def seeded_func(*args, seed=seed, **kwargs):
+        if seed is not None:
+            np.random.seed(seed)
+        return func(*args, **kwargs)
+    return seeded_func
+
+
+
+def split_range_sample(split_range, p=None):
+    """
+    Return a value sampled randomly from the given split range.
+    
+    Parameters
+    ----------
+    split_range : list(tuple)
+        A list of [min, max) tuples defining ranges from which to sample.
+        The sample will be drawn uniformly from the ranges, as if they were
+        contiguous. For example, on an input of [(0,1),(2,3)], a sampled
+        value of 1.5 will return 2.5.
+        
+    p : list(float)
+        If given, should be a list the same length as split_range, and
+        contains the probability of sampling from each range. p will be
+        normalized before use. Defaults to None.
+        
+    Returns
+    -------
+    A value sampled uniformly from the given split range.
+    """
+    if p is None:
+        total = sum([val[1] - val[0] for val in split_range])
+
+        sample = uniform(0, total)
+
+        for bottom, top in split_range:
+            if sample < top - bottom:
+                return bottom + sample
+            sample -= top - bottom
+
+        raise ValueError('ERROR: split_range_sample tried to return value ' +
+                         'outside of any range.')
+    
+    # Here, p is used to sample a range first
+    p = p / np.sum(p)
+    index = choice(range(len(split_range)), p=p)
```

Comment:
You know...I found this a bit difficult to understand. Instead of writing the whole `if p is None:` clause above, why not always just use np.random.choice (which will default to even probabilities if `p=None`) i.e. I'm suggesting that you simply delete lines 63 - 75 inclusive, and insert an `if p is not None` to set p before the line selecting the index (currently line 78).

That way it's totally clear you just do a two step process:
1. Select a range
1. Uniformly sample the range